pkg/util/informermanager/federatedinformermanager.go

```diff
@@ -0,0 +1,339 @@
+	clientGetter ClusterClientGetter,
+	ftcInformer fedcorev1a1informers.FederatedTypeConfigInformer,
+	clusterInformer fedcorev1a1informers.FederatedClusterInformer,
+) FederatedInformerManager {
+	manager := &federatedInformerManager{
+		lock:                        sync.RWMutex{},
+		started:                     false,
+		clientGetter:                clientGetter,
+		ftcInformer:                 ftcInformer,
+		clusterInformer:             clusterInformer,
+		eventHandlerGenerators:      []*EventHandlerGenerator{},
+		clusterEventHandlers:         []*ClusterEventHandler{},
+		clients:                     map[string]dynamic.Interface{},
+		connectionMap:               map[string][]byte{},
+		informerManagers:            map[string]InformerManager{},
+		informerManagersCancelFuncs: map[string]context.CancelFunc{},
+		queue:                       workqueue.NewRateLimitingQueue(workqueue.DefaultItemBasedRateLimiter()),
+	}
+
+	clusterInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			cluster := obj.(*fedcorev1a1.FederatedCluster)
+			return util.IsClusterJoined(&cluster.Status)
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { manager.enqueue(obj) },
+			UpdateFunc: func(_ interface{}, obj interface{}) { manager.enqueue(obj) },
+			DeleteFunc: func(obj interface{}) { manager.enqueue(obj) },
+		},
+	})
+
+	ftcInformer.Informer()
+
+	return manager
+}
+
+func (m *federatedInformerManager) enqueue(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		klog.Error(err, "federated-informer-manager: Failed to enqueue FederatedCluster")
+		return
+	}
+	m.queue.Add(key)
+}
+
+func (m *federatedInformerManager) worker(ctx context.Context) {
+	key, shutdown := m.queue.Get()
+	if shutdown {
+		return
+	}
+	defer m.queue.Done(key)
+
+	logger := klog.FromContext(ctx)
+
+	_, name, err := cache.SplitMetaNamespaceKey(key.(string))
+	if err != nil {
+		logger.Error(err, "Failed to process FederatedCluster")
+		return
+	}
+
+	logger = logger.WithValues("cluster", name)
+	ctx = klog.NewContext(ctx, logger)
+
+	cluster, err := m.clusterInformer.Lister().Get(name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		logger.Error(err, "Failed to get FederatedCluster from lister, will retry")
+		m.queue.AddRateLimited(key)
+		return
+	}
+	if apierrors.IsNotFound(err) || !util.IsClusterJoined(&cluster.Status) {
+		if err := m.processClusterDeletion(ctx, name); err != nil {
+			logger.Error(err, "Failed to process FederatedCluster, will retry")
+			m.queue.AddRateLimited(key)
+			return
+		}
+		return
+	}
+
+	err, needReenqueue := m.processCluster(ctx, cluster)
+	if err != nil {
+		if needReenqueue {
+			logger.Error(err, "Failed to process FederatedCluster, will retry")
+		} else {
+			logger.Error(err, "Failed to process FederatedCluster")
+		}
+	}
+	if needReenqueue {
+		m.queue.AddRateLimited(key)
+	}
```

Comment:
We need to call `queue.Forget` on all code paths where we need to reset the backoff, including but not limited to here.

I would suggest moving the code from `_, name, err := cache.SplitMetaNamespaceKey(key.(string))` to `err, needReenqueue := m.processCluster(ctx, cluster)` into `processCluster`, and choose whether to backoff or reset backoff based on its return value.

Reply:
I would rather not bloat `processCluster` even more and also make it responsible for getting the cluster from the lister + checking if the cluster exists. I will simply add `queue.Forget` at the necessary places in the `worker` method.